### ISSUE TITLE
Weird behavior of training for convolution and maxpool layers

### ISSUE THREAD

**Issue by e3oroush** [keras-team/keras#9848]
I'm using Keras 2.15 with tensorflow backend. I have faced a weird behavior from keras training `Sequential model .fit()` function for this past three days. I defined a simple CNN network with one layer `Conv2d` and one `max-pooling` followed by two fully connected layers for the classification. I conducted an experiment to check the stability of the training in keras. I froze all the random values by setting feed to some constant value and also initializing all layer weights with constant values.  
What I expect from keras training module is to give me exactly identical losses, all layer weights through epochs of the training not almost but exactly. I've included my scripts of experiments in this post.  
For reproducing my experiments, you should run this code 3 times, first set variable `initing=True` to initialize all weights and store it to the disk. Then, set `initing=False` and run it two times; This will store all weights of the network in `state%d.ktf` and gradients to the file `state%d.grads`. You can compare these values with function `comapre_list_weights(path1,path2)`.  
## Notes
* I conducted the same experiment with tensorflow and pytorch and got the expected results, so there is nothing wrong with my network or expectations
* If I remove the `Max-pooling` layer the problem would be resolved!

```python
import os
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
# make sure use no gpu
os.environ["CUDA_VISIBLE_DEVICES"] = ""
from time import time
import numpy as np
np.random.seed(22)
import keras.backend as K
from tensorflow import set_random_seed
set_random_seed(22)
from keras.layers import Dense,Conv2D,MaxPooling2D,Flatten
from keras.models import Sequential
from keras.optimizers import Adam
from keras.callbacks import Callback
import pickle

initing=True
n_classes=1
BATCHSIZE=1
n=32
nb_channel=3
nb_data=2
x_train=np.array([np.random.normal(0,0.1,size=(nb_channel,n,n)),np.random.normal(0.5,0.1,size=(nb_channel,n,n))])
y_train=np.array([0,1])

def simple_cnn_network():
    model = Sequential()   
    model.add(Conv2D(50, kernel_size=(3, 3), padding='same', 
    activation='relu',data_format='channels_first',
                 input_shape=(nb_channel,n,n),name='conv1'))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2),data_format='channels_first',name='maxpool1'))
    
    model.add(Flatten())
    model.add(Dense(512, activation='relu',name='fc1'))
    model.add(Dense(n_classes, activation='sigmoid',name='output'))
    return model

def keras_weird_beh():
    weight_path="keras_cnn_weights.h5"
    # callback class for storing history of weights and their grads after each batch
    class ParameterHistory(Callback):
        states=nb_data*NumEpoches*[None]
        gradients=nb_data*NumEpoches*[None]
        epoch=0
        def on_epoch_begin(self, epoch, logs=None):
            self.epoch=epoch
        def on_batch_begin(self,batch, logs=None):
            pass
        def on_batch_end(self, batch, logs=None):
            weights=self.model.get_weights()
            self.states[batch+(self.epoch)*nb_data]=weights
            inputs = [[x_train[batch]], # X
                        [1], # sample weights
                        [[y_train[batch]]], # y
                        0 # learning phase in TEST mode
                     ]
            self.gradients[batch+self.epoch*nb_data]=get_gradients(inputs)
            
    model=simple_cnn_network()
    if not initing:
        model.load_weights(weight_path)
        print("Loaded model from disk")

    model.compile(
        loss = "binary_crossentropy",
        optimizer = Adam(lr=1e-3,beta_1=0.9,beta_2=0.999,epsilon=1e-8,amsgrad=False))
    model.summary()
    # the following is to compute gradient of layers
    weights=[tensor for tensor in model.trainable_weights if model.get_layer(tensor.name.split('/')[0]).trainable]
    optimizer=model.optimizer
    grads=optimizer.get_gradients(model.total_loss,weights)
    input_tensors = [model.inputs[0], # input data
            model.sample_weights[0], # how much to weight each sample by
            model.targets[0], # labels
            K.learning_phase(), # train or test mode
            ]
    get_gradients = K.function(inputs=input_tensors, outputs=grads)
    
    cb=ParameterHistory()
    model.fit(x_train,
          y_train,
          batch_size=BATCHSIZE,
          epochs=NumEpoches,
          verbose=1,
          shuffle=False,
          callbacks=[cb])

    if initing:
        model.save(weight_path)
        print("Saved model to disk")
    else:
        states=cb.states
        gradients=cb.gradients
        np.random.seed(int(time()))
        path='state%d.ktf'%int(np.random.randint(100))
        print('saving state tp %s'%path)
        with open(path, 'wb') as fhandler:
            pickle.dump(states, fhandler)
        with open(path.split('.')[0]+'.grads', 'wb') as fhandler:
            pickle.dump(gradients,fhandler)

def comapre_list_weights(path1,path2):
    with open(path1, 'rb') as fhandler:
        w1=pickle.load(fhandler)
    with open(path2,'rb') as fhandler:
        w2=pickle.load(fhandler)
    names=['weight','bias']
    for i, w_1 in enumerate(w1):
        diff=0
        for j,w__1 in enumerate(w_1):
            diff=np.abs(w__1-w2[i][j]).sum()
            print('diff at epoch %d batch %d layer %d %s is %.6f'%(i//2,int(i%2),j//2,names[int(j%2)],diff))




if initing:
    NumEpoches = 1
else:
    NumEpoches = 5

if __name__=='__main__':
    keras_weird_beh()

```
